core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```
@@ -0,0 +1,123 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep{
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  public static void main(String[] args) {
+    if (args.length != 3) {
+      throw new IllegalArgumentException("USAGE: JavaGrep regex rootPath outFile");
+    }
+
+    //Default logger configuration
+    BasicConfigurator.configure();
+    JavaGrepImp javaGrepImp = new JavaGrepImp();
+    javaGrepImp.setRegex(args[0]);
+    javaGrepImp.setRootPath(args[1]);
+    javaGrepImp.setOutFile(args[2]);
+
+    try {
+      javaGrepImp.process();
+    }catch (Exception ex){
+      javaGrepImp.logger.error(ex.getMessage(), ex);
+    }
+  }
+
+  @Override
+  public void process() throws IOException {
+    List<String> matchedLines = new ArrayList<>();
+    for (File file : listFiles(getRootPath())){
+      for (String line : readLines(file)){
+        if (containsPattern(line)){
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    List<File> files = new ArrayList<>();
+    File folder = new File(rootDir);
+    File[] allFiles = folder.listFiles();
+    for (File f: allFiles){
+      if(f.isFile()){
+        files.add(f);
+      }
+    }
+    return files;
+  }
+
+  @Override
+  public List<String> readLines(File inputFile) {
+    Path path = Paths.get(inputFile.toURI());
+    List<String> lines = new ArrayList<>();
+    try {
+      lines = Files.readAllLines(path);
+    }catch (IOException e){
+      logger.error("Unable to read lines", e);
```

you might want to throw a runtime exception to stop the program